src/common/transformations/src/transformations/symbolic_transformations/symbol_optimization.cpp

```diff
@@ -218,25 +219,106 @@ void optimize_value_usage(ov::Output<ov::Node>& output, STS_map& symbol_shape_so
     }
 }
 
-void save_shape_sources(const ov::Output<ov::Node>& output, STS_map& symbol_shape_source) {
-    for (const auto& d : output.get_partial_shape()) {
-        if (d.is_static())
-            continue;
-        auto symbol = d.get_symbol();
-        if (symbol == nullptr)
-            continue;
-        if (symbol_shape_source.count(symbol))
+std::vector<std::shared_ptr<ov::Node>> topological_order(const std::shared_ptr<ov::Model>& m) {
+    // step 1: split model into parameter related and parameter non-related ops
+    const std::string op_depends_on_parameter = "topological_sort_op_depends_on";
+    // values: true - parameter dependent; false otherwise
+    for (const auto& op : m->get_ordered_ops()) {
+        if (ov::as_type_ptr<ov::op::v0::Parameter>(op)) {
+            op->get_rt_info()[op_depends_on_parameter] = true;
+        } else if (ov::as_type_ptr<ov::op::v0::Constant>(op) || ov::as_type_ptr<ov::op::v0::ShapeOf>(op) ||
+                   ov::as_type_ptr<ov::op::v3::ShapeOf>(op) ||
+                   std::dynamic_pointer_cast<ov::op::util::VariableExtension>(op)) {
+            op->get_rt_info()[op_depends_on_parameter] = false;
+        } else {  // deduce op type from inputs
+            const auto& inputs = op->input_values();
+            op->get_rt_info()[op_depends_on_parameter] =
+                std::any_of(inputs.begin(), inputs.end(), [=](const ov::Output<ov::Node>& input) {
+                    return input.get_node_shared_ptr()->get_rt_info()[op_depends_on_parameter].as<bool>();
+                });
+        }
+    }
+    // step 2: starting from Result -- assign weight to ops:
+    //      if parameter dependant, weights is maximum of output indices plus one
+    //      else weights is maximum of output indices
+    // this step doesn't assign weights to all the ops, this is intentional and will be used in the following step
+    const std::string weight_rt_info_name = "topological_sort_weight";
+    auto reverse_order = m->get_ordered_ops();
+    std::reverse(reverse_order.begin(), reverse_order.end());
+    for (const auto& op : reverse_order) {
```

Comment:
Maybe is possible to not use `std::reverse` but replace `for` loop using reverse iterator from model's ordered ops?